tutorials/Configuration.md
Installing and Configuring CAT
==============================

Packages required prior to installing CAT
-----------------------------------------
The CAT generates installers for numerous operating systems. Consequently, many console tools need to be installed for all the profile generators to work correctly. The prerequisites are:

* Apache2 Web Server
* MySQL/MariaDB Server (we recommend MariaDB and regularly test against versions MariaDB 5.5 and MySQL 5.7)
* PHP 8.0.0 or higher; on Ubuntu also the "php8.0-intl" package
* Required PHP extensions: "gettext", "openssl", "PECL:IMagick", "GD" and "MySQL"
* Optional PHP extensions: "GeoIP" (deprecated v1 GeoIP API)
* simpleSAMLphp version 2.0.0 or higher (included as a composer dependency)
* NSIS 3.00 or higher - either as a native Linux binary or on Wine
* zip
* wpa_supplicant (eapol_test utility, with support for all contemporary EAP methods compiled in - esp. EAP-FAST)
* OpenSSL

Configuring the required prerequisite packages
----------------------------------------------
Here are some extra configuration hints for these packages:

* Operating System

	language display needs the corresponding locales to be installed (check config/config-template.php for the exact list of locales that CAT can support right now)
	
	for better entropy for cryptographic operations (important particularly for Managed IdP deployments), be sure to have "haveged" service installed and running
	
* MySQL / MariaDB

	the timezone should be set to UTC in my.cnf
	
* Apache

	the Directory for installer downloads (configurable, defaults to web/downloads/ ) needs to have "AllowOverrides FileInfo" set

	the directories under web/ need to be accessible from DocumentRoot
	
	the CAT log dir (configurable, defaults to /var/log/CAT/ ) needs to be accessible for writing
	
	for general server hardening, the following configuration tokens should be set: 
           - ServerSignature Off
           - ServerTokens ProductOnly
           - TraceEnable Off

        the following header directives should be configured:
           - Header always set x-xss-protection "1; mode=block"
           - Header always set x-frame-options "SAMEORIGIN"
           - Header always set x-content-type-options "nosniff"
	
	if you want to use client certificates for administrative user authentication, be sure set a sufficiently large SSL Renegotiation Buffer size (e.g. SSLRenegBufferSize 10486000 for 10 MB max. upload size)
	
	There are custom error pages for 404 etc. in web/404.php etc. If you want to use them, set ErrorDocument 404 /.../404.php etc. in your Apache virtual host config.
* PHP

	for general server hardening, the following option should be set in php.ini: "expose_php 0"

	for cookie security, the following options should be set in php.ini: "session.cookie_httponly 1" and "session.cookie_secure 1"

        the extension php-gmp is needed when enabling Silverbullet functionality

	To send mails, PHPMailer v6 needs to be on your system. If your package manager does not provide it, you can download the tarball from GitHub and place it in the directory core/PHPMailer

* simpleSAMLphp

	configure it as a service provider, authenticating towards an IdP of your choice. Attribute mapping is defined in config.php
* NSIS

	needs to have the plug-in "nsArray"
	
	"makensis" needs to be configured in the config/config.php file and executable
* GeoIP

	API Version 1:

	best install as a system package or use instructions from http://dev.maxmind.com/geoip/downloadable#PHP-7 or http://php.net/manual/en/geoip.installation.php
	
	download GeoLiteCity and GeoLiteCityv6 databases from http://geolite.maxmind.com/download/geoip/database/GeoLiteCity.dat.gz and http://geolite.maxmind.com/download/geoip/database/GeoLiteCityv6.dat.gz
	
	unzip them and copy into GeoIP directory changing names to GeoCity.dat and GeoCityv6.dat (the directory is /usr/share/GeoIP or something similar, GeoIP will display an error message if the database is missing and you will be able to guess the required location).
	
	arrange for downloads each month (databases are updated on the first Thursday of each month)
	
	API Version 2:
	
	TBD

Installing CAT
--------------
1.  unpack the distribution
1A. if you use a clone of the Git repo instead, remember to "git submodule --init --recursive" at least for the GEANTlink repo in devices/ms/Files/ (there are more submodules referenced in core/ which you may already have on your system, you should double-check)
2.  create the config/config-master.php file from the supplied template config-master-template.php
3.  create the config/config-diagnostics.php and/or config/config-confassistant.php files as needed in the same manner
4.  create the devices/Devices.php file from the supplied template
5.  on a MySQL/MariaDB server, create the databases as per the schema definition in schema/schema.sql
6.  make sure that you can connect to that database
7.  make sure that the var/installer_cache, var/silverbullet and var/tmp directories exists and are writable to the Apache web server user
8.  make sure that simpleSAMLphp is installed
9.  make sure that simpleSAMLphp openid module is enabled and google (or any IdP of your choice) is uncommented in authsources
10.  create a symlink web/external/jquery/jquery.js pointing to the most current version of jQuery. When installing CAT from composer, the most recent version will be available in vendor/components/jquery/jquery.min.js (the version shipped inside the distribution may not be the most current one!)
11. using your browser, check if the main interface is running (web subdirectory)
12. if so, go to the master management page to have your system prerequisites checked (web/admin/112365365321.php)

Configuring CAT
---------------
After creating config-*.php as above, adapt it to your needs and the realities on your server. In particular, pay attention to the following:

* reference the autoloader of your simpleSAMLphp installation correctly in config-master.php
* enter the connection details to the database(s)

The device configuration file is in devices/Devices.php. There is a template file devices-template.php - you can simply copy it to have a Devices.php. Unless you want to disable specific device modules, or have custom ways to digitally sign installers, it is not necessary to change this file.

After logging in with simpleSAMLphp, you should enable protection of the superadmin page. This can be done by editing the config-master "SUPERADMINS" array - remove everything and instead add only the unique identifiers that are supposed to have access to that page.

Similarly, you may want to make yourself the federation administrator for at least one federation. You can do so by adding that privilege to your unique identifier with the following SQL statement in the USER database (in this example, for "LU" -> Luxembourg):

```INSERT INTO user_options (user_id, option_name, option_value) VALUES("<your unique ID>","user:fedadmin","LU");```

Additional configuration for Managed IdP
----------------------------------------
* edit the shell script utils/ocsp_update-template.sh to make it point to the location of the OCSP responder.
* make sure SSH access to the OCSP responder works without user interaction using the configured access details.
* set up cron job to run the shell script utils/ocsp_update-template.sh regularly (e.g. once per hour)

Customisation / Look and Feel
-----------------------------
CAT ships with a default look-and-feel for the eduroam consortium. You can adapt most of its appearance to your local needs by changing colours and images. The sources for the images are scattered around the source tree. Here is a list of LOGOs to edit:

* web/resources/images/consortium_logo.png (website main logo)
* web/resources/images/gradient-bg.png (gradient top-down for the adverising 'film roll'
* web/resources/images/screenshots/* (sample installers on 'film roll' - supplied ones carry digital sig from TERENA and eduroam logo; might not be appropriate for your use
* web/favicon.ico (website favicon)
* devices/ms/Files/eduroam_150.bmp (logo to embed in Microsoft installers)
* devices/ms/Files/eduroam32.ico (window icon for Microsoft installers)
* devices/ms/Files/cat_bg.bmp (background for front page on windows installers)

The colours are all configured in the web/resources/css/ directory.

* The default eduroam colour number 1 is #BCD7E8 (light blue). Change it to your own colours as needed.
* The default eduroam colour number 2 is #0A698E (dark blue). Change it to your own colours as needed.


web/apidoc/files/core/Options.php.txt
<?php
/*
 * *****************************************************************************
 * Contributions to this work were made on behalf of the GÉANT project, a 
 * project that has received funding from the European Union’s Framework 
 * Programme 7 under Grant Agreements No. 238875 (GN3) and No. 605243 (GN3plus),
 * Horizon 2020 research and innovation programme under Grant Agreements No. 
 * 691567 (GN4-1) and No. 731122 (GN4-2).
 * On behalf of the aforementioned projects, GEANT Association is the sole owner
 * of the copyright in all material which was developed by a member of the GÉANT
 * project. GÉANT Vereniging (Association) is registered with the Chamber of 
 * Commerce in Amsterdam with registration number 40535155 and operates in the 
 * UK as a branch of GÉANT Vereniging.
 * 
 * Registered office: Hoekenrode 3, 1102BR Amsterdam, The Netherlands. 
 * UK branch address: City House, 126-130 Hills Road, Cambridge CB2 1PQ, UK
 *
 * License: see the web/copyright.inc.php file in the file structure or
 *          <base_url>/copyright.php after deploying the software
 */

/**
 * This file contains some convenience functions around option handling.
 *
 * @author Stefan Winter <stefan.winter@restena.lu>
 *
 * @package Developer
 */
/**
 * necessary includes
 */

namespace core;

use \Exception;

/**
 * The Options class contains convenience functions around option handling. It is implemented as a singleton to prevent
 * excessive DB requests; its content never changes during a script run.
 *
 * @author Stefan Winter <stefan.winter@restena.lu>
 */
class Options {

    /**
     * database which this class queries by default
     * 
     * @var string
     */
    private static $databaseType = "INST";

    /**
     * The (only) instance of this class
     * 
     * @var Options
     */
    private static $instance;

    /**
     * Our access to logging facilities
     * 
     * @var \core\common\Logging
     */
    private $loggerInstance;

    /**
     * This private variable contains the list of all known options and their properties (i.e. flags).
     * 
     * @var array all known options
     */
    private $typeDb;

    const TYPECODE_STRING = "string";
    const TYPECODE_INTEGER = "integer";
    const TYPECODE_TEXT = "text";
    const TYPECODE_BOOLEAN = "boolean";
    const TYPECODE_FILE = "file";
    const TYPECODE_COORDINATES = "coordinates";

    /**
     * Returns the handle to the (only) instance of this class.
     * 
     * @return Options
     */
    public static function instance() {
        if (!isset(self::$instance)) {
            $className = __CLASS__;
            self::$instance = new $className;
        }
        return self::$instance;
    }

    /**
     * Prevent cloning - this is a singleton.
     * 
     * @return void
     */
    public function __clone() {
        trigger_error('Cloning not allowed for singleton classes.', E_USER_ERROR);
    }

    /**
     *  Option class constructor; retrieves information about the known options from the database.
     */
    private function __construct() {
        $this->typeDb = [];
        $this->loggerInstance = new \core\common\Logging();
        $this->loggerInstance->debug(3, "--- BEGIN constructing Options instance ---\n");
        $handle = DBConnection::handle(self::$databaseType);
        $options = $handle->exec("SELECT name,type,flag from profile_option_dict ORDER BY name");
        // SELECT -> resource, not boolean
        while ($optionDataQuery = mysqli_fetch_object(/** @scrutinizer ignore-type */ $options)) {
            $this->typeDb[$optionDataQuery->name] = ["type" => $optionDataQuery->type, "flag" => $optionDataQuery->flag];
        }
        $this->typeDb["general:logo_url"] = ["type" => "string", "flag" => NULL];
        $this->typeDb["eap:ca_url"] = ["type" => "string", "flag" => NULL];
        $this->typeDb["internal:country"] = ["type" => "string", "flag" => NULL];
        $this->typeDb["internal:profile_count"] = ["type" => "integer", "flag" => NULL];
        $this->typeDb["internal:checkuser_outer"] = ["type" => "boolean", "flag" => NULL];
        $this->typeDb["internal:checkuser_value"] = ["type" => "string", "flag" => NULL];
        $this->typeDb["internal:verify_userinput_suffix"] = ["type" => "boolean", "flag" => NULL];
        $this->typeDb["internal:hint_userinput_suffix"] = ["type" => "boolean", "flag" => NULL];
        $this->typeDb["internal:realm"] = ["type" => "string", "flag" => NULL];
        $this->typeDb["internal:use_anon_outer"] = ["type" => "boolean", "flag" => NULL];
        $this->typeDb["internal:anon_local_value"] = ["type" => "string", "flag" => NULL];        
        $this->loggerInstance->debug(3, "--- END constructing Options instance ---\n");
    }

    /**
     * This function lists all known options. If called with the optional parameter $className, only options of that class are
     * returned, otherwise the full set of all known attributes.
     * 
     * @assert ("user") == Array("user:email","user:fedadmin","user:realname")
     * 
     * @param string $className optionally specifies the class of options to be listed (class is the part of the option name before the : sign)
     * @return array of options
     */
    public function availableOptions($className = 0) {
        $tempArray = [];
        $this->loggerInstance->debug(3, "CLASSNAME IS $className\n");

        foreach (array_keys($this->typeDb) as $name) {
            if ($className === 0) {
                $tempArray[] = $name;
            } elseif (preg_match('/^' . $className . ':/', $name) > 0) {
                $tempArray[] = $name;
            }
        }
        $returnArray = $tempArray;
        // remove silverbullet-specific options if this deployment is not SB
        foreach ($tempArray as $key => $val) {
            if (( \config\Main::FUNCTIONALITY_LOCATIONS['CONFASSISTANT_SILVERBULLET'] != 'LOCAL') && (preg_match('/^fed:silverbullet/', $val) > 0)) {
                unset($returnArray[$key]);
            }
            if (( \config\Main::FUNCTIONALITY_LOCATIONS['CONFASSISTANT_RADIUS'] != 'LOCAL') && (preg_match('/^fed:minted_ca_file/', $val) > 0)) {
                unset($returnArray[$key]);
            }
        }
        return $returnArray;
    }

    /** This function returns the properties of a given attribute name. This currently means it returns its type and its flag field ("ML").
     *
     * @assert ("general:instname") == Array("type"=>"string", "flag"=>"ML")
     * @assert ("profile:production") == Array("type"=>"boolean", "flag"=>NULL)
     * 
     * @param string $optionname Name of the option whose properties are to be retrieved.
     * @return array properties of the attribute
     * @throws Exception
     */
    public function optionType($optionname) {
        if (isset($this->typeDb[$optionname])) {
            return $this->typeDb[$optionname];
        }
        throw new Exception("Metadata about an option was requested, but the option name does not exist in the system: " . htmlentities($optionname));
    }

    /**
     * This function is mostly useless. It takes an (unvetted) string, sees if
     * it is a valid option name, and then returns the array key of the typeDb
     * instead of the unvetted string. This makes Scrutinizer happy.
     * 
     * @param string $unvettedName the input name
     * @return string the name echoed back, but from trusted source
     * @throws Exception
     */
    public function assertValidOptionName($unvettedName) {
        $listOfOptions = array_keys($this->typeDb);
        foreach ($listOfOptions as $name) {
            if ($name == $unvettedName) {
                return $name;
            }
        }
        throw new Exception("Unknown option name encountered.");
    }

}



Changes
Changes in 1.2
================
- [FEATURE #1] UserAPI redone. Instead of the "id" as a common argument we now use
               meaningful names, like idp, profile, device etc.
               To get the new behaviour you need to set api_version argument to 2.
- [FEATURE #2] added createTemporaryDirectory to the Helper to avoind using the came code in several places
- [FEATURE #3] configuration tests rebuilt and extended
- [FEATURE #4] realm checks are saved in DB and results shown on federation 
               overview page
- [FEATURE #5] federation customisation: name, logo, custom invitation texts and
               more
- [FEATURE #6] deprecated NSISArray has been replaced with nsArry
- [FEATURE #7] Support for UTF-8 installer has been added (this requires nsis v3)

Configuration parameter changes
-------------------------------

Changes in 1.1.2
================
- [BUGFIX #1]  in radiust_tests translation of sme messages was not activated 
               and they appeared in English
- [BUGFIX #2]  in Windows installers the profile name was not sent trough code page
               conversion (reported by Daniele Albrizio)
- [BUGFIX #3]  fix regression: federation operators see their national stats
               again
- [BUGFIX #4]  In UserAPI generateInstaller returned a link to the actual file location and not the proper API call
- [BUGFIX #5]  Move installer cache away from the web tree
- [BUGFIX #6]  in Linux module fix multiline dispplay of messages in terminal mode, the installer can be run with sh, better bash location
- [BUGFIX #7]  In Linux installer there was no support for SSID removal
- [BUGFIX #8]  Fix language changing in realm checks
- [BUGFIX #9]  Fix RADIUS tests for self-signed server certs
- [BUGFIX #10] UDP tests running in parallel could fail due to the same Calling-Station-Id
- [BUGFIX #11] Server certificate subject printing failed if multiple CNs were present
- [BUGFIX #12] In "live login" RADIUS tests there was a mixup with passed identities
- [BUGFIX #13] In RADIUS tests the temporary directory was not deleted
- [BUGFIX #14] EAP-PWD implementation renewed and in the same time a problem with loading profiles on 64-bit systems resolved
- [BUGFIX #15] fix unathenticated access to installers marked as non-production
- [FEATURE #1] "About CAT" now displays a matrix of supported devices and EAP
               type combinations
- [FEATURE #2] RADIUSTests now use the first tab to display overview information from all tests
- [FEATURE #3] Linux module now supports Python3
- [FEATURE #4] Linux module verifies server certificates bases on altSubjectName
- [FEATURE #5] Linux module asks the user if in case of Network Manager installation failure
               it should save a wpa_supplicant file

Configuration parameter changes
-------------------------------
[NEW]           Config::$APPEARANCE['support-contact'] replaces Config::$APPEARANCE['admin-mail']


Changes in 1.1.1
================
- [BUGFIX #1]  device-level redirect not working on a selected device
- [BUGFIX #2]  mobileconfig proxy setting corrected
- [BUGFIX #3]  in RADIUS tests the contents of results box needs to be cleared
               the moment that a rerun of tests is called
- [BUGFIX #4]  in basic.php fix a problem resulting from geolocation failure
- [BUGFIX #5]  in roll.php - set the translation domain to het translations in
- [BUGFIX #6]  hide hidden devices in basic.php
- [BUGFIX #7]  fix language selection problem in admin pages
- [BUGFIX #8]  fix problem caused by installer generation delays
- [BUGFIX #9] in some sutuations the footer was obsuring parts of the text
- [BUGFIX #12] make clear that we support Mac OS X 10.7 and above, not 10.6
- [UPDATE #1]  in devices.php added the pointer to eduroamCAT on Google play
- [UPDATE #2]  speed up IdP listing
- [UPDATE #3]  update of system names in roll.php
- [FEATURE #1] in "About", show the hostnames which need to be whitelisted in
               captive portals for CAT to work. This required two new config
               parameters, see below.
- [FEATURE #2] ArnesLink for TTLS in Windows 7 and Vista modules
- [FEATURE #3] added ChromeOS as beta-quality device module
- [FEATURE #6] allow configuration of an abuse mail contact. See below.
- [FEATURE #7] add separate device module for iOS 5+6 to fix proxy regression


Configuration parameter changes
-------------------------------
[NEW]           Config::$APPEARANCE['webcert_CRLDP']
[NEW]           Config::$APPEARANCE['webcert_OCSP']
[NEW]           Config::$APPEARANCE['abuse-mail']

Changes in 1.1
==============
- [BUGFIX #1]  installer cache problem has been fixed by adding a mime field to the downloads 
               table in the database, hence a schema change
- [BUGFIX #2]  Linux installer - passwords containing single quote broke
               fix suggested by Brian Epstein <bepstein@ias.edu>
- [BUGFIX #3]  Linux installer - added tkip in group protocol
               to support wpa/wpa2 mixed mode
- [BUGFIX #4]  Linux installer - the name of local installation directory
               changed from .eduroam to .cat_installer
- [BUGFIX #5]  Cleared a potential vulnarability in cat_info (item 9 from the 
               audit)
- [BUGFIX #6]  Cleared a potential vulnarability in radius_tests.php (item 11 
               from the audit)
- [BUGFIX #7]  Cleared a potential vulnarability in cat_js.php (item 13 from the
               audit)
- [BUGFIX #8]  Cleared a potential vulnarability in index.php (item 20 from the
               audit)
- [BUGFIX #9]  Added server hardening best practices in the Configuration
               tutorial (closing items D2 and D4 of the audit)
- [BUGFIX #10] Input sanitisation in productheader() improved (closes item 7
               from the source code audit)
- [BUGFIX #11] all other source code audit issues fixed
- [BUGFIX #12] removed XP support and Win Vista and 7 support for TTLS due to
               open licensing questions on a particular third-party helper

core/Options.php
<?php

/*
 * *****************************************************************************
 * Contributions to this work were made on behalf of the GÉANT project, a 
 * project that has received funding from the European Union’s Framework 
 * Programme 7 under Grant Agreements No. 238875 (GN3) and No. 605243 (GN3plus),
 * Horizon 2020 research and innovation programme under Grant Agreements No. 
 * 691567 (GN4-1) and No. 731122 (GN4-2).
 * On behalf of the aforementioned projects, GEANT Association is the sole owner
 * of the copyright in all material which was developed by a member of the GÉANT
 * project. GÉANT Vereniging (Association) is registered with the Chamber of 
 * Commerce in Amsterdam with registration number 40535155 and operates in the 
 * UK as a branch of GÉANT Vereniging.
 * 
 * Registered office: Hoekenrode 3, 1102BR Amsterdam, The Netherlands. 
 * UK branch address: City House, 126-130 Hills Road, Cambridge CB2 1PQ, UK
 *
 * License: see the web/copyright.inc.php file in the file structure or
 *          <base_url>/copyright.php after deploying the software
 */

/**
 * This file contains some convenience functions around option handling.
 *
 * @author Stefan Winter <stefan.winter@restena.lu>
 *
 * @package Developer
 */
/**
 * necessary includes
 */

namespace core;

use \Exception;

/**
 * The Options class contains convenience functions around option handling. It is implemented as a singleton to prevent
 * excessive DB requests; its content never changes during a script run.
 *
 * @author Stefan Winter <stefan.winter@restena.lu>
 */
class Options
{

    /**
     * database which this class queries by default
     * 
     * @var string
     */
    private static $databaseType = "INST";

    /**
     * The (only) instance of this class
     * 
     * @var Options
     */
    private static $instance;

    /**
     * Our access to logging facilities
     * 
     * @var \core\common\Logging
     */
    private $loggerInstance;

    /**
     * This private variable contains the list of all known options and their properties (i.e. flags).
     * 
     * @var array all known options
     */
    private $typeDb;

    const TYPECODE_STRING = "string";
    const TYPECODE_ENUM_LIST = "enum_openroaming";
    const TYPECODE_INTEGER = "integer";
    const TYPECODE_TEXT = "text";
    const TYPECODE_BOOLEAN = "boolean";
    const TYPECODE_FILE = "file";
    const TYPECODE_COORDINATES = "coordinates";

    public const LEVEL_METHOD = "Method";
    public const LEVEL_PROFILE = "Profile";
    public const LEVEL_IDP = "IdP";
    public const LEVEL_FED = "FED";
    public const LEVEL_USER = "User";

    /**
     * Returns the handle to the (only) instance of this class.
     * 
     * @return Options
     */
    public static function instance()
    {
        if (!isset(self::$instance)) {
            $className = __CLASS__;
            self::$instance = new $className;
        }
        return self::$instance;
    }

    /**
     * Prevent cloning - this is a singleton.
     * 
     * @return void
     */
    public function __clone()
    {
        trigger_error('Cloning not allowed for singleton classes.', E_USER_ERROR);
    }

    /**
     *  Option class constructor; retrieves information about the known options from the database.
     */
    private function __construct()
    {
        $this->typeDb = [];
        $this->loggerInstance = new \core\common\Logging();
        $uiElements = new \web\lib\admin\UIElements();
        $this->loggerInstance->debug(4, "--- BEGIN constructing Options instance ---\n");
        $handle = DBConnection::handle(self::$databaseType);
        $options = $handle->exec("SELECT name,type,flag, multivalued from profile_option_dict ORDER BY name");
        // SELECT -> resource, not boolean
        while ($optionDataQuery = mysqli_fetch_object(/** @scrutinizer ignore-type */ $options)) {
            $optionValue = $uiElements->displayName($optionDataQuery->name, true);
            if ($optionValue != false) {
                $this->typeDb[$optionDataQuery->name] = ["type" => $optionDataQuery->type,
                    "flag" => $optionDataQuery->flag, "multivalued" => $optionDataQuery->multivalued];
            }
        }
        $this->typeDb["general:logo_url"] = ["type" => "string", "flag" => NULL, "multivalued" => 0];
        $this->typeDb["eap:ca_url"] = ["type" => "string", "flag" => NULL, 'multivalued' => 1];
        $this->typeDb["internal:country"] = ["type" => "string", "flag" => NULL, "multivalued" => 0];
        $this->typeDb["internal:profile_count"] = ["type" => "integer", "flag" => NULL, "multivalued" => 0];
        $this->typeDb["internal:checkuser_outer"] = ["type" => "boolean", "flag" => NULL, "multivalued" => 0];
        $this->typeDb["internal:checkuser_value"] = ["type" => "string", "flag" => NULL, "multivalued" => 0];
        $this->typeDb["internal:verify_userinput_suffix"] = ["type" => "boolean", "flag" => NULL, "multivalued" => 0];
        $this->typeDb["internal:hint_userinput_suffix"] = ["type" => "boolean", "flag" => NULL, "multivalued" => 0];
        $this->typeDb["internal:realm"] = ["type" => "string", "flag" => NULL, "multivalued" => 0];
        $this->typeDb["internal:use_anon_outer"] = ["type" => "boolean", "flag" => NULL, "multivalued" => 0];
        $this->typeDb["internal:anon_local_value"] = ["type" => "string", "flag" => NULL, "multivalued" => 0];
        $this->loggerInstance->debug(4, "--- END constructing Options instance ---\n");
    }

    /**
     * This function lists all known options. If called with the optional parameter $className, only options of that class are
     * returned, otherwise the full set of all known attributes.
     * 
     * @assert ("user") == Array("user:email","user:fedadmin","user:realname")
     * 
     * @param string $className optionally specifies the class of options to be listed (class is the part of the option name before the : sign)
     * @param int $showMultiFlag optionally show the single/multi valued flag for the option
     * @return array of options
     */
    public function availableOptions($className = 0, $showMultiFlag=0)
    {
        $showMultiFlag=0;
        $tempArray = [];
        foreach ($this->typeDb as $name=>$values) {
            if (!\core\CAT::hostedServicesEnabled() && (preg_match('/^fed:silverbullet/', $name) > 0)) {
                continue;
            }
            if (!\core\CAT::radiusProfilesEnabled() && (preg_match('/^fed:minted_ca_file/', $name) > 0)) {
                continue;
            }
            if ($className === 0) {
                $tempArray[] = $showMultiFlag === 1 ? [$name, $values['multivalued']] : $name;
            } elseif (preg_match('/^' . $className . ':/', $name) > 0) {
                $tempArray[] = $showMultiFlag === 1 ? [$name, $values['multivalued']] : $name;
            }
        }
        return $tempArray;
    }

    /** This function returns the properties of a given attribute name. This currently means it returns its type and its flag field ("ML").
     *
     * @assert ("general:instname") == Array("type"=>"string", "flag"=>"ML")
     * @assert ("profile:production") == Array("type"=>"boolean", "flag"=>NULL)
     * 
     * @param string $optionname Name of the option whose properties are to be retrieved.
     * @return array properties of the attribute
     * @throws Exception
     */
    public function optionType($optionname)
    {
        if (isset($this->typeDb[$optionname])) {
            return $this->typeDb[$optionname];
        }
        return false;
    }

    /**
     * This function is mostly useless. It takes an (unvetted) string, sees if
     * it is a valid option name, and then returns the array key of the typeDb
     * instead of the unvetted string. This makes Scrutinizer happy.
     * 
     * @param string $unvettedName the input name
     * @return string the name echoed back, but from trusted source
     * @throws Exception
     */
    public function assertValidOptionName($unvettedName)
    {
        $listOfOptions = array_keys($this->typeDb);
        foreach ($listOfOptions as $name) {
            if ($name == $unvettedName) {
                return $name;
            }
        }
        throw new Exception("Unknown option name encountered.");
    }
}
